Ensure packages can be built with TypeScript strict flag

devversion opened the pull request:
Implemented as part of FW-1265. See individual commits for description of changes.

devversion committed:
fix(ivy): incorrect type definition for ɵɵdefineComponent

Currently the `ɵɵdefineComponent` method has incorrect type
definitions the `directives` and `pipes` metadata property.

The incorrect types allow developers to pass in already instantiated
`DirectiveDef` or `ComponentDef` objects. This can cause unexpected
failures because the definition internally only expects `Type` objects
and now incorrectly tries to read the `ngDirectiveDef` or `ngComponentDef`
of existing definitions.

This issue has been unveiled by enabling the strict function parameter
types flag, where the directive definitions are determined from each array
element in the `directives` or `pipes` property (which can throw).

devversion committed:
fix(ivy): semantic module check incorrectly handles nested arrays

In View Engine, developers can pass bootstrap and entry components
as nested arrays. e.g.

```ts
export const MyOtherEntryComponents = [A, B, C]

@NgModule({
  entryComponents: [MyComp, MyOtherEntryComponents]
})
```

Currently using nested arrays for these properties causes
unexpected errors to be reported in Ivy since the semantic
NgModule checks aren't properly recursing into the nested
entry/bootstrap components. This issue has been unveiled by
enabling the strict function parameter checks.

gkalpak commented on packages/compiler/src/render3/view/template.ts:
```diff
@@ -1609,7 +1611,7 @@ export class BindingScope implements LocalResolver {
 
   generateSharedContextVar(retrievalLevel: number) {
-    const lhs = o.variable(CONTEXT_NAME + this.freshReferenceName());
+    const lhs: o.ReadVarExpr = o.variable(CONTEXT_NAME + this.freshReferenceName());
```
Why this change (since `o.variable` [already returns][1] `o.ReadVarExpr`)?

[1]: https://github.com/angular/angular/blob/c4ab32f664ee8f4619893db508fc2f0cf0f132f3/packages/compiler/src/output/output_ast.ts#L1490-L1493

devversion replied:
I did that to make it clear that shared context variables should be of type `o.ReadVarExpr`, but looking at it again it seems weird. Removed.

gkalpak commented on packages/tsconfig-build-strict.json:
```diff
@@ -0,0 +1,10 @@
+{
+  "extends": "./tsconfig-build.json",
+  "compilerOptions": {
+    "strict": true
```
OOC, have you also tried `--noImplicitReturns`, `--noUnusedLocals`, `--noUnusedParameters`?
(I assume there will be a ton of violations :grin:)

devversion replied:
Nope not yet :smile: We should definitely look into that, but probably out of scope for the `--strict` change for now :grin:

gkalpak reviewed (commented):
:tada:

devversion commented on packages/core/src/di/reflective_provider.ts:
```diff
@@ -180,10 +180,11 @@ export function mergeResolvedReflectiveProviders(
   providers.forEach(b => {
     if (b instanceof Type) {
-      res.push({provide: b, useClass: b});
+      res.push({ provide: b, useClass: b } as NormalizedProvider);
```
Yes. I had the same feeling. `NormalizedProvider` seems to be incorrect since it extends from `TypeProvider`. Let's look into this in a separate PR/story.

alexeagle commented on packages/BUILD.bazel:
```diff
@@ -23,6 +23,12 @@ ts_config(
 
+ts_config(
+    name = "tsconfig-build-strict",
+    src = "tsconfig-build-strict.json",
```
I'm trying to remove all the extra tsconfig files. What would be needed to avoid adding a new one? need a refactoring tool to automatically correct each `--strict` option as we enable it? We are quite good at doing that in google3 where we've always had a single tsconfig for the whole monorepo

devversion replied:
You mean something like a genrule to generate the modified tsconfig dynamically? Should we handle this outside of this PR? (seems like a general question of how we manage these configs)

devversion commented on packages/core/src/debug/debug_node.ts:
```diff
@@ -411,7 +411,7 @@ class DebugElement__POST_R3__ extends DebugNode__POST_R3__ implements DebugEleme
   queryAll(predicate: Predicate<DebugElement>): DebugElement[] {
     const matches: DebugElement[] = [];
-    _queryAllR3(this, predicate, matches, true);
+    _queryAllR3(this, predicate as Predicate<DebugNode>, matches, true);
```
Very good question. This is actually a bit odd because the predicate wants to be called with `DebugElement` instances, while the `_queryAllR3` function only *calls it* with `DebugNode`. 

Now with `strictFunctionTypes`.. this fails because technically something could use `DebugElement` specific logic in the predicate, while only a `DebugNode` got passed. See [here](https://www.typescriptlang.org/play/#src=interface%20Base%20%7B%0D%0A%20%20myProp%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Ainterface%20Derived%20extends%20Base%20%7B%0D%0A%20%20derivedProp%3A%20string%3B%0D%0A%7D%0D%0A%0D%0Afunction%20test(cb%3A%20(c%3A%20Base)%20%3D%3E%20void)%20%7B%20%7D%0D%0A%0D%0Atest((a%3A%20Base)%20%3D%3E%20%7B%20%7D)%3B%0D%0Atest((a%3A%20Derived)%20%3D%3E%20%7B%7D)%3B)).

I'll add another commit that fixes this under-the-hood so that we don't need this "hacky" cast.

gkalpak replied:
I didn't realize `_queryAllR3()` was eventually calling `predicate()` (with a `DebugNode`). Makes sense now.

devversion commented on packages/core/test/acceptance/ng_module_spec.ts:
```diff
@@ -0,0 +1,65 @@
+    });
+  });
+
+});
```
We actually only needed one call to `createComponent` to ensure that the test module is created. I'll either add a comment explaining this or going the safe and less-confusing way of calling `createComponent` for both components.

devversion commented on packages/core/src/util/decorators.ts:
```diff
@@ -49,10 +49,11 @@ export function makeDecorator<T>(
   const metaCtor = makeMetadataCtor(props);
 
-  function DecoratorFactory(...args: any[]): (cls: Type<T>) => any {
+  function DecoratorFactory(this: void|typeof DecoratorFactory, ...args: any[]): (cls: Type<T>) =>
```
I like that more. done!

devversion commented on packages/core/src/render3/di.ts:
```diff
@@ -557,7 +557,7 @@ export function getNodeInjectable(
     setTNodeAndViewData(tNode, lData);
     try {
-      value = lData[index] = factory.factory(null, tData, lData, tNode);
+      value = lData[index] = factory.factory(undefined, tData, lData, tNode);
```
Very good question. The reason for this change is that the canonical factories are typed using the `FactoryFn` interface:

https://github.com/angular/angular/blob/0ee09cdd7e5f076355775322ae7d21de8f23c40a/packages/core/src/render3/interfaces/definition.ts#L45-L50

The `NodeInjectorFactory` currently calls these factories for directives with `null` as the first parameter. This is incorrect and could cause unexpected behavior. We could have solved this differently by wrapping the directive factory functions like that: `(_: null) => directiveFactory()`, but  this results in a few function declaration for each directive. Additionally it looks like the initial goal was to type the `NodeInjectorFactory#factory` equally to the `FactoryFn` (to avoid these wrappings).. AFAICT that's why we even pass around `null` (and now `undefined` :grin:)

devversion committed:
refactor(common): ensure compatibility with typescript strict flag

As part of FW-1265, the `@angular/common` package is made compatible
with the TypeScript `--strict` flag. Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

devversion committed:
refactor(upgrade): ensure compatibility with typescript strict flag

As part of FW-1265, the `@angular/upgrade` package is made compatible
with the TypeScript `--strict` flag. Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

devversion committed:
refactor(platform-browser): compatibility with typescript strict flag

As part of FW-1265, the `@angular/platform-browser` package is made compatible
with the TypeScript `--strict` flag. Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

devversion committed:
refactor(router): compatibility with typescript strict flag

As part of FW-1265, the `@angular/router` package is made compatible
with the TypeScript `--strict` flag. Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

devversion committed:
refactor(forms): ensure compatibility with typescript strict flag

As part of FW-1265, the `@angular/forms` package is made compatible
with the TypeScript `--strict` flag. Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

devversion committed:
refactor(service-worker): ensure compatibility with typescript strict flag

As part of FW-1265, the `@angular/service-worker` package is made compatible
with the TypeScript `--strict` flag. Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

petebacondarwin commented on packages/core/src/render3/styling/host_instructions_queue.ts:
```diff
@@ -61,7 +61,7 @@ function findNextInsertionIndex(buffer: HostInstructionsQueue, priority: number)
  */
-export function flushQueue(context: StylingContext): void {
+export function flushQueue(this: any, context: StylingContext): void {
   const buffer = context[StylingIndex.HostInstructionsQueue];
```
I see both `this: any` and `this: unknown`.
Is there a difference? If not then we should be consistent and just use one.

devversion replied:
Thanks Pete. Yes a few instances in this PR need to be switched to `unknown`. Note that also a few need to remain as `any` if the context is clear (but not type-able). e.g. `createDynamicClass` which uses `Object.create`.

devversion committed:
refactor(core): ensure compatibility with typescript strict flag

As part of FW-1265, the `@angular/core` package is made compatible
with the TypeScript `--strict` flag. This already unveiled a few bugs,
so the strictness flag seems to help with increasing the overall code health.

Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

devversion committed:
refactor(compiler): ensure compatibility with typescript strict flag

As part of FW-1265, the `@angular/compiler` package is made compatible
with the TypeScript `--strict` flag. This already unveiled a few bugs,
so the strictness flag seems to help with increasing the overall code health.

Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

devversion committed:
refactor: fix remaining typescript strict flag failures

Fixes the remaining TypeScript --strict flag failures for source
files which are not part of any specific release package.

petebacondarwin commented on packages/compiler/src/output/output_interpreter.ts:
```diff
@@ -79,7 +79,7 @@ function createDynamicClass(
   const ctorParamNames = _classStmt.constructorMethod.params.map(param => param.name);
   // Note: use `function` instead of arrow function to capture `this`
-  const ctor = function(...args: any[]) {
+  const ctor = function(this: any, ...args: any[]) {
```
Actually I note that oin general using `any` hides potential bugs...
For example the `_ExecutionContext` constructor accepts an `instance` parameter which is `any` but on one occasion we are passing `null` here, then later in the code we access this value unguarded.  We could consider something like this, which at least makes the assertion that we believe `instance` is not `null` explicit:

```diff
diff --git a/packages/compiler/src/output/output_interpreter.ts b/packages/compiler/src/output/output_interpreter.ts
index 7f63f7bb6c..344fe37c06 100644
--- a/packages/compiler/src/output/output_interpreter.ts
+++ b/packages/compiler/src/output/output_interpreter.ts
@@ -38,8 +38,8 @@ class _ExecutionContext {
   exports: string[] = [];
 
   constructor(
-      public parent: _ExecutionContext|null, public instance: any, public className: string|null,
-      public vars: Map<string, any>) {}
+      public parent: _ExecutionContext|null, public instance: Object|null,
+      public className: string|null, public vars: Map<string, any>) {}
 
   createChildWihtLocalVars(): _ExecutionContext {
     return new _ExecutionContext(this, this.instance, this.className, new Map<string, any>());
@@ -79,9 +79,9 @@ function createDynamicClass(
 
   const ctorParamNames = _classStmt.constructorMethod.params.map(param => param.name);
   // Note: use `function` instead of arrow function to capture `this`
-  const ctor = function(...args: any[]) {
+  const ctor = function(this: Object, ...args: any[]) {
     const instanceCtx = new _ExecutionContext(_ctx, this, _classStmt.name, _ctx.vars);
-    _classStmt.fields.forEach((field) => { this[field.name] = undefined; });
+    _classStmt.fields.forEach((field) => { (this as any)[field.name] = undefined; });
     _executeFunctionStatements(
         ctorParamNames, args, _classStmt.constructorMethod.body, instanceCtx, _visitor);
   };
@@ -125,7 +125,7 @@ class StatementInterpreter implements o.StatementVisitor, o.ExpressionVisitor {
     if (ast.builtin != null) {
       switch (ast.builtin) {
         case o.BuiltinVar.Super:
-          return ctx.instance.__proto__;
+          return Object.getPrototypeOf(ctx.instance);
         case o.BuiltinVar.This:
           return ctx.instance;
         case o.BuiltinVar.CatchError:
@@ -188,7 +188,7 @@ class StatementInterpreter implements o.StatementVisitor, o.ExpressionVisitor {
     const args = this.visitAllExpressions(stmt.args, ctx);
     const fnExpr = stmt.fn;
     if (fnExpr instanceof o.ReadVarExpr && fnExpr.builtin === o.BuiltinVar.Super) {
-      ctx.instance.constructor.prototype.constructor.apply(ctx.instance, args);
+      ctx.instance !.constructor.prototype.constructor.apply(ctx.instance, args);
       return null;
     } else {
       const fn = stmt.fn.visitExpression(this, ctx);
```

devversion replied:
Totally agree. Even though this seems like a general loose type issue within the `output_interpreter` and is not necessarily in scope for this PR, I'm changing it to what you proposed in order to avoid that we need to type `this` with `any`.

devversion committed:
refactor: fix typescript strict flag failures in all tests

Fixes all TypeScript failures caused by enabling the `--strict`
flag for test source files. We also want to enable the strict
options for tests as the strictness enforcement improves the
overall codehealth, unveiled common issues and additionally it
allows us to enable `strict` in the `tsconfig.json` that is picked
up by IDE's.

devversion committed:
fix(platform-browser): debug element query predicates not compatible with strictFunctionTypes

Currently developers can use the `By` class to construct common
`DebugElement` query predicates. e.g. `By.directive(MyDirective)`.

The `directive()` and `all()` predicates are currently returning
a predicate that works for `DebugElement` nodes. This return type
is too strict since the predicate is not specific to `DebugElement`
instances and can also apply to `DebugNode` instances.

Meaning that developers are currently able to use the `directive()`
predicate when using `queryAllNodes()`. This is a common practice
but will break when the project is compiled with TypeScript's
`--strictFunctionTypes` flag as the `DebugElement` predicate type
is not assignable to predicates for `DebugNode`. In order to make
these predicates usable with `--strictFuntionTypes` enabled, we
adjust the predicate type to reflect what is actually needed for
evaluation of the predicate.

alxhub commented on packages/core/testing/src/r3_test_bed_compiler.ts:
```diff
@@ -508,20 +508,23 @@ export class R3TestBedCompiler {
+          if (!descriptor) {
+            // Delete operations are generally undesirable since they have performance implications
+            // on objects they were applied to. In this particular case, situations where this code
+            // is
```
Some formatting weirdness going on here :)

devversion replied:
Good catch :smile: Done!

alxhub reviewed (approved):
😲 👏 

Amazing work, sir. Thank you!!

devversion committed:
refactor: ensure zone.js can be built with typescript strict flag

As part of FW-1265, the `zone.js` package is made compatible
with the TypeScript `--strict` flag. Read more about the strict flag [here](https://www.typescriptlang.org/docs/handbook/compiler-options.html)

devversion commented on packages/platform-browser/src/dom/debug/by.ts:
```diff
@@ -44,14 +44,14 @@ export class By {
    */
-  static directive(type: Type<any>): Predicate<DebugElement> {
-    return (debugElement) => debugElement.providerTokens !.indexOf(type) !== -1;
+  static directive(type: Type<any>): Predicate<DebugNode> {
```
These return type changes can seem like a breaking change, but `Predicate<DebugNode>` is also assignable to `Predicate<DebugElement>` it should not be breaking.

It just _fixes_ the type so that `debugEl.query(By.directive(..))` also works for consumer applications using `strictFunctionTypes`. Here is a playground demo: [see here](https://www.typescriptlang.org/play/?noImplicitAny=false&strictFunctionTypes=false&noImplicitReturns=false&alwaysStrict=false#code/LAKA9AVBoAQTBBGAjA9qgNgUwIYDsBaANxwwFcsATGAMzLwGMAXAS1TxlSKwCcYcYJclgA0MAA6oAzlJbIMATxgtG5SioDmMBuyZYAHk2V4aqHgFscrdrHg8sGnD3V4tTABZXBpCgHIpEtIs1hwq-BxOPDgKAHS2tjAAAuJk8iwMCOIstmCgKno8NDgMWDAACvbqDFZYADwAKgB8MADeMAAUQhQAXDD1AJS9aJi4eADcMAC+oHl4BUUlMAAiWMhkGgByqJSlLbAwB3jbWBWo4r1STDyaY6DTILPzxaUraxoAotjmWHMwBnp4SgBV7rLY7Vr7A5YDCnc4wS7XVy3ED3UB0RghfgMEriJhSEEfL4-JgVKjpGrtfS9UlVGq1AmfLDfOaNfqte7o5hsCLYrC4-GrUHHGnkvSU6mVUV1AlgrCs9kzEDFHF4hlEuYi6pi9p4XpqpnEtkAXmaVwo-WRyr5qsFhINGslWqw7R1ettsuNpp45uRoCt-JlwsdFNdy1tjOZTE9MDNWDZYDA8PcqDIGGoRRYGBgAHdgu54Vd0kY8Kmswx3FgGABrAJhHR4KRkb58HDicQYUXcmJ9dwsWsBAziSt6SgAQhg9lItBwmbI9j9vID7qDZKdLt1YbeHpgJpj3rjYyAA)

JiaLiPassion reviewed (approved):
Thanks!

IgorMinar reviewed (approved):
thanks Paul!!

mhevery commented:
http://test/OCL:258796685:BASE:258796726:1563470812584:47aac9ff
